src/util/recentlyEditedTree.js

```diff
@@ -0,0 +1,115 @@
+import _ from 'lodash'
+import { timestamp } from './timestamp'
+import { pathToContext } from './pathToContext'
+import { timeDifference } from '../util.js'
+
+const editTimeMax = 30 // time diff limit in second for replacing descendants by ancestor
+
+const findTreeDeepestSubcontext = (tree, path) => {
+  if (path.length === 0) return { node: {}, path: [] }
+  const availableNode = _.get(tree, path)
+  if (availableNode) return { node: availableNode, path }
+  const pathIndex = path.findIndex((value, index) => _.get(tree, (path.slice(0, path.length - index))))
+  return pathIndex > -1 ? { node: _.get(tree, path.slice(0, path.length - pathIndex)), path: path.slice(0, path.length - pathIndex) } : {}
+}
+
+export const findTreeDescendants = (tree, startingPath, leafNodes = []) => {
+  const node = _.get(tree, startingPath)
+  if (!node) return []
+  if (node.leaf) leafNodes.push({ ...node }) // eslint-disable-line fp/no-mutating-methods
+  else Object.keys(node).forEach(child => findTreeDescendants(tree, [...startingPath, child], leafNodes))
+  return leafNodes
+}
+
+const findClosestAncestorWithMultipleChildren = (tree, context) => {
+  const contextIndex = context.findIndex((value, index) => {
+    const node = _.get(tree, (context.slice(0, context.length - index)))
+    return !node.leaf && Object.keys(node).length > 1
+  })
+  return contextIndex > -1 ? { node: _.get(tree, context.slice(0, context.length - contextIndex + 1)), path: context.slice(0, context.length - contextIndex + 1) } : {}
```

Comment:
I recommend placing each `key: value` pair on a separate line to help readability with all but the smallest object declarations.